### DISCUSSION TITLE
Are Reference Maps supported?

### DISCUSSION THREAD

**Discussion by hasezoey** [Automattic/mongoose#10311] [Q&A] [answered]
basically my question is "are reference maps supported" and when yes, is this the right way or does the `ref` option need to be placed somewhere else

```ts
// NodeJS: 16.1.0
// MongoDB: 4.2-bionic (Docker)
import * as mongoose from "mongoose"; // mongoose@5.12.10
import { inspect } from "util";

const DummySchema = new mongoose.Schema({
  dummy: String
});
const TestSchema = new mongoose.Schema({
  testprop: {
    type: Map,
    of: mongoose.Schema.Types.ObjectId,
    ref: "dummy"
  }
});

const DummyModel = mongoose.model("dummy", DummySchema);
const TestModel = mongoose.model("test", TestSchema);

(async () => {
  await mongoose.connect(`mongodb://localhost:27017/`, { useNewUrlParser: true, dbName: "verifyMASTER", useCreateIndex: true, useUnifiedTopology: true });

  console.log("schema path", TestModel.schema.path("testprop"));

  const dummy_doc1 = await DummyModel.create({ dummy: "hello1" });
  const dummy_doc2 = await DummyModel.create({ dummy: "hello2" });

  const test_doc1 = await TestModel.create({ testprop: [["hello1", dummy_doc1], ["hello2", dummy_doc2]] });

  const found_doc1 = await TestModel.findById(test_doc1).orFail().exec();

  console.log("found", inspect(found_doc1, false, 5));

  await found_doc1.populate("testprop.$*");

  console.log("populated", inspect(found_doc1, false, 5));

  await mongoose.disconnect();
})();
```

PS: thanks to #9359 i already know that it is supported to have references inside maps with an extra property, but my question is when the ObjectId is directly the type of the map

**Comment by vkarpov15** [comment, marked as answer]
At first glance this code looks right. Are you getting an error or some sort of unexpected behavior?

**Reply by hasezoey** [reply]
no error and it "works", just populating does nothing, but i also dont know if i even did it correctly for an map

**Reply by vkarpov15** [reply]
Ok cool, I'll move this to an issue then

**Reply by vkarpov15** [reply]
See https://github.com/Automattic/mongoose/issues/10329

**Reply by hasezoey** [reply]
thanks, i was not really sure if this was an issue / bug or if i just didnt know how to use it properly

**Reply by vkarpov15** [reply]
I took a look and the below works:

```javascript
const TestSchema = new mongoose.Schema({
  testprop: {
    type: Map,
    of: { type: mongoose.Schema.Types.ObjectId, ref: 'dummy' }
  }
});
```

It looks like we just need `ref` to bubble down into the map.

**Reply by hasezoey** [reply]
thanks for this info, i wasnt quite sure if this would create an object map
and now the question for me is, is this still considered a bug / issue to fix, or just something with `SchemaTypeOptions`?

Edit:
maybe the original was phrased weird, let me re-try:
now my question is, is this gonna be automatically done (from the "top" level(original code example) into the level down(your code example)), or does this need to be done manually (like via `SchemaTypeOptions` mapping or something else like adding an extra case for maps and the option `ref`)
